@here/harp-mapview/lib/ClipPlanesEvaluator.ts

```diff
@@ -0,0 +1,524 @@
+            // Store camera position tmpVectors[0] and reference it with p.
+            const p = this.m_tmpVectors[0].copy(camera.position);
+            p.addScaledVector(fwdRot, Math.sqrt(d * d - r * r));
+            farPlane = p.sub(camera.position).dot(fwd);
+            const bias = 2000; // TODO: generalize.
+            nearPlane = Math.max(this.nearMin, projection.groundDistance(camera.position) - bias);
+        } else {
+            const groundDistance = projection.groundDistance(camera.position);
+            nearPlane = Math.max(this.nearMin, groundDistance * this.nearMultiplier);
+            // Will be already clamped to minFar due to clamping above.
+            farPlane = nearPlane * this.nearFarMultiplier + this.farOffset;
+        }
+        return { near: nearPlane, far: farPlane };
+    }
+}
+
+/**
+ * Simple camera clip planes evaluator that computes near plance based on camera to ground distance.
+ *
+ * This evaluator supports both planar and spherical projections, although it behaviour is
+ * slightly different in each case. General algorithm setups near plane and far plane close
+ * to ground level, but taking into account some predefined margins.
+ * @note This simple evaluator does behave correctly when camera is always pointing downwards the
+ * ground surface (top view) along surface normal, but does not preserve correct planes when using
+ * modifying camera pitch (tilt) angle.
+ */
+export class AltitudeBasedClipPlanesEvaluator implements ClipPlanesEvaluator {
+    protected static readonly GROUND_NORMAL_INV_PLANAR_PROJ = new THREE.Vector3(0, 0, -1);
+
+    protected m_tmpVectors: THREE.Vector3[] = [
+        new THREE.Vector3(),
+        new THREE.Vector3(),
+        new THREE.Vector3()
+    ];
+    protected m_tmpQuaternion: THREE.Quaternion = new THREE.Quaternion();
+
+    /**
+     * Allows to setup near/far offsets (margins) from the ground as also minimum near value.
+     * All c-tor parameters should take positive numbers or zeros (with exception of [[nearMin]]).
+     * It is stronlgy recommended to set some reasonable margins/offset for near and far planes
+     * from ground surface to avoid flickering. Keep in mind that sum of the margins:
+     * [[nearMargin]] + [[farMargin]] can not be bigger then maximum offset between a planes
+     * assumed - [[planesMaxDistance]].
+     * @note [[planesMaxDistance]] does not limit far plane when spherical projection is in use,
+     * the algorithm used estimates distance to point on tangent where line from camera touches
+     * the sphere horizont.
+     * @param nearMargin defines near plane offset from the ground in the surface normal direction.
+     * @param farMargin defines far plane offset from the ground surface, looking downwards.
+     * @param nearMin minimum allowable near plance distance from camera.
+     * @param planesMaxDistance maximum offset between near/far planes, allows to limit viewing
```

Comment:
does this work when fully zoomed out? Wouldn't this risk hiding the sphere / plane?

Reply:
That is true, but it is up to user if he want's to setup this, you may always pass maximum possible number, but for certain scenarios it may be quite useful. Imagine that you allow to tilt camera at 89 degrees, your view would then spread to infinity, so we need to have some maximum limit at least to save performance.